### PR TITLE
feat(optimizer): Implement `ProjectJoin` rule

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Merge projects into joins's output indices where project's expressions are all `InputRef`s

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml [hunk 1]
```diff
@@ -554,7 +552,7 @@
       LogicalJoin { type: LeftSemi, on: (t2.y = (t3.y + t2.y)) AND (t2.y = t2.y) AND (t2.x = t3.x), output: [t2.x] }
         LogicalScan { table: t2, columns: [t2.x, t2.y] }
         LogicalProject { exprs: [t3.x, t2.y, (t3.y + t2.y)] }
-          LogicalJoin { type: Inner, on: true, output: all }
+          LogicalJoin { type: Inner, on: true, output: [t3.x, t2.y, t3.y] }
```
Comment (jon-chuang) [inline src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:555]: Weird, what was changed here?

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/subquery_expr_correlated.yaml:555]: It could be due to an intermediate project being merged.

**Comment by chenzl25** [comment]
IIRC, physical optimization likes `toBatch` or `toStream` always create project for `LogicalJoin` with a non-trivial output_indices. Can we benefit from this rule?

**Comment by st1page** [comment]
> IIRC, physical optimization likes `toBatch` or `toStream` always create project for `LogicalJoin` with a non-trivial output_indices. Can we benefit from this rule?

IIRC, we have support output_indices in executor, what the `non-trivial` means

**Comment by chenzl25** [comment]
> 

In this case trivial means identity.

**Comment by chenzl25** [comment]
> > 
> 
> In this case trivial means identity.

Oh, I just only focus on the pull_filter part which always create a project. No question Now.

**Review by xxchan — APPROVED** [review]
related issue https://github.com/singularity-data/risingwave/issues/1922

Will we also need to similarly merge simple project into output indices of other nodes? 🤔

**Comment by jon-chuang** [comment]
> Will we also need to similarly merge simple project into output indices of other nodes? 

Possibly. One of these is scan. I'm not sure if its necessary to do this as it may already be facilitated by prune_cols?

**Pull request merged**